File: leetcode/src/main/ruby/zig_zag.rb
# @param {String} s
# @param {Integer} num_rows
# @return {String}
def convert(s, num_rows)
    return s if num_rows == 1

    len = s.length
    cycle = 2 * (num_rows - 1)

    sb = ''
    for i in 0...num_rows
        j = i
        while j < len
            sb << s[j]
            rj = cycle + j - 2 * i
            sb << s[rj] if (i != 0 && i != num_rows - 1 && rj < len)
            j += cycle
        end
    end
    sb
end

File: leetcode/src/main/golang/zig_zag.go
func convert(s string, numRows int) string {
    if numRows <= 1 {
        return s
    }

    slen := len(s)
    cycle := 2 * (numRows - 1)

    sb := make([]byte, slen)
    idx := 0
    for i := 0; i < numRows; i++ {
        for j := i; j < slen; j += cycle {
            sb[idx] = s[j]
            idx++

            rj := cycle + j - 2 * i
            if i != 0 && i != numRows - 1 && rj < slen {
                sb[idx] = s[rj]
                idx++
            }
        }
    }

    return string(sb)
}
